File: Recursion/TailExample.c
#include <stdio.h>

int fib(int n, int a, int b) {
  if (n == 0) {
    return a;
  } else if (n == 1) {
    return b;
  } else {
    return fib(n - 1, b, a + b);
  }
}
int fibonacci(int n) {
  return fib(n, 0, 1);
}

int fac(int n, int acc) {
  if (n == 0) {
    return acc;
  } else {
    return fac(n - 1, n * acc);
  }
}
int factorial(int n) {
  return fac(n, 1);
}

int main(void) {
  for (int i = 0; i <= 10; i++) {
    printf("fibonacci(%d) = %d\n", i, fibonacci(i));
  }
  for (int i = 0; i <= 10; i++) {
    printf("factorial(%d) = %d\n", i, factorial(i));
  }
  return 0;
}

File: rekurencja-tail-example-f/TailExample.fs
let rec fib n a b =
  match n with
  | 0 -> a
  | 1 -> b
  | _ -> fib (n - 1) b (a + b)
let fibonacci n = fib n 0 1

let rec fac n acc =
  match n with
  | 0 -> acc
  | _ -> fac (n - 1) (n * acc)
let factorial n = fac n 1 

for i = 1 to 10 do
  printf "fibonacci(%d) = %d\n" i (fibonacci i)

for i = 1 to 10 do
  printf "factorial(%d) = %d\n" i (factorial i)